src/tactic/linarith/elimination.lean

```diff
@@ -149,21 +147,25 @@ and does not appear in the sum.
 Computing the sum of the two comparisons is easy; the complicated details lie in tracking the
 additional fields of `pcomp`.
 * The historical set `pcomp.history` of `c1 + c2` is the union of the two historical sets.
-* We recompute the variables that appear in `c1 + c2` from the newly created `linexp`,
-  since some may have been implicitly eliminated.
+* `vars` is the union of `c1.vars` and `c2.vars`.
 * The effectively eliminated variables of `c1 + c2` are the union of the two effective sets,
   with `elim_var` inserted.
-* The implicitly eliminated variables of `c1 + c2` are those that appear in at least one of
-  `c1.vars` and `c2.vars` but not in `(c1 + c2).vars`, excluding `elim_var`.
+* The implicitly eliminated variables of `c1 + c2` are those that appear in
+  `vars` but not `c.vars` or `effective`.
+(Note that the description of the implicitly eliminated variables of `c1 + c2` in the algorithm
+described in Section 6 of https://doi.org/10.1016/B978-0-444-88771-9.50019-2 seems to be wrong:
+that says it should be `(c1.implicit.union c2.implicit).sdiff explicit`.
+Since the implicitly eliminated sets start off empty for the assumption,
+this formula would leave them always empty.)
 -/
 meta def pcomp.add (c1 c2 : pcomp) (elim_var : ℕ) : pcomp :=
 let c := c1.c.add c2.c,
     src := c1.src.add c2.src,
     history := c1.history.union c2.history,
-    vars := native.rb_set.of_list c.vars,
+    vars := c1.vars.union c2.vars,
     effective := (c1.effective.union c2.effective).insert elim_var,
-    implicit := ((c1.vars.union c2.vars).sdiff vars).erase elim_var in
-⟨c, src, history, effective, implicit, vars⟩
+    implicit := (vars.sdiff (rb_set.of_list c.vars)).sdiff effective in
+⟨c, src, history, vars, effective, implicit⟩
```

Comment:
Ouch, was this an anonymous field ordering bug?

Reply:
I don't think so, it looks like the fields were reordered in this PR as well. (Probably that change can be reverted, it was the result of removing and then re-adding the field in a different place.)

Reply:
Yes, this was incidental.

Reply:
Do you prefer this new order, or should I push a commit to restore the old one in the interest of a small diff?

Reply:
If you could restore the old one that would be great. Sorry I haven't got back to this. :-(

Reply:
Done :)